File: 1/2020.10.18/17.Letter Combinations of a Phone Number/python/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        def DFS(i, tmp):
            # terminator
            if len(tmp) == len(digits):
                res.append(tmp)
                return
            # process current level
            for c in d[digits[i]]:
                # drill down
                DFS(i + 1, tmp + c)
                # reset states
        
        if not digits:
            return []
        d = {}
        d['2'] = 'abc'
        d['3'] = 'def'
        d['4'] = 'ghi'
        d['5'] = 'jkl'
        d['6'] = 'mno'
        d['7'] = 'pqrs'
        d['8'] = 'tuv'
        d['9'] = 'wxyz'
        res = []
        DFS(0, '')
        return res

File: 1/2020.10.18/17.Letter Combinations of a Phone Number/java/solution.java
class Solution {

    Map<Character, String> map = new HashMap<>();
    List<String> res = new ArrayList<>();

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return res;
        }
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");
        DFS(digits, 0, "");
        return res;
    }

    private void DFS(String digits, int start, String tmp) {
        // terminator
        if (start == digits.length()) {
            res.add(tmp);
            return;
        }
        // process current level
        for (char c : map.get(digits.charAt(start)).toCharArray()) {
            // drill down
            DFS(digits, start + 1, tmp + String.valueOf(c));
        }
    }
}
